text/0000-multiarch-builders-and-package.md

```diff
@@ -0,0 +1,944 @@
+
+- What is the expected outcome?
+
+`pack buildpack package` and `pack builder create` commands will be updated in a way that they will handle the creation of multi-arch OCI images
+
+# What it is
+[what-it-is]: #what-it-is
+
+The end-users for this proposal are **Buildpack authors**, we expect to improve their user experience when creating multi-architecture buildpacks and builders as follows
+
+## Multi-arch example
+
+Let's suppose a **Buildpack author** has a `buildpack.toml` created with a content similar to this one:
+
+```toml
+# Buildpack API version
+api = "0.12"
+
+# Buildpack ID and metadata
+[buildpack]
+  id = "examples/my-multiarch-buildpack"
+  version = "0.0.1"
+
+# List of targets operating systems, architectures and versions
+
+[[targets]]
+os = "linux"
+arch = "amd64"
+
+[[targets]]
+os = "linux"
+arch = "arm64"
+
+[[targets]]
+os = "windows"
+arch = "amd64"
+
+[[targets.distributions]]
+name = "windows"
+versions = ["10.0.20348.1970"]
+```
+
+And organizes the binaries according to their os/arch with a structure similar to this one:
+
+```bash
+my-multiarch-buildpack
+├── buildpack.toml
+├── linux-amd64
+│         └── bin
+│             ├── build
+│             └── detect
+├── linux-arm64
+│         └── bin
+│             ├── build
+│             └── detect
+└── windows-amd64
+    └── 10.0.20348.1970
+        └── bin
+            ├── build.bat
+            └── detect.bat
+```
+
+Now `pack` will be able to package them separately for each os/arch family, following our [guide](https://buildpacks.io/docs/buildpack-author-guide/package-a-buildpack/) 
+we will need to create a `package.toml` file.
+
+The `package.toml` file will be similar to the one describe in our documentation:
+
+```toml
+[buildpack]
+uri = "examples/my-multiarch-buildpack"
+# OR a .tgz with the previous folder structure
+uri = "my-multiarch-buildpack.tgz"
+```
+
+But in this case, we actually **remove** the [platform](https://buildpacks.io/docs/reference/config/package-config/) section because it will be taken from the `buildpack.toml`.
+
+Packaging a multi-arch buildpack will require the output to be **publish** to a registry or **saved on disk** in OCI layout format.
+
+```bash
+pack buildpack package my-buildpack --config ./package.toml --publish --multi-arch
+# Or
+pack buildpack package my-buildpack.cnb --config ./package.toml --format file --multi-arch
+```
+
+In these cases each `target` entry corresponds to a different buildpack image that is exported into an [image index](https://github.com/opencontainers/image-spec/blob/master/image-index.md)
+
+### Package a multi-arch Builder
+
+In case of packing a **Builder**, we assume the following premises: 
+
+1. Buildpack authors migrated their `builder.toml` and [remove the stack](https://github.com/buildpacks/pack/issues/1303) concept.
```

Comment:
I'm concerned about this. If we make the change in #1303, will we still be able to support stacks? We have a lot of users in production. We need to support both for a while so we can give folks plenty of time to migrate.

If this requires removing and stopping support for stacks then I suspect that this will hold us up for a while on being able to use this functionality.

Is there any way we could change this to something like `Buildpack authors migrated `builder.toml` and have added the entries to `builder.toml` for the new run/build image format.`. So that we could have both for some period of time as we migrate?

Reply:
Let me double check this, I need the targets specify in the `builder.toml`. If the users still specify stacks for compatibility I think we can workaround it.

Reply:
I think for the purposes of this RFC any `[[stacks]]` in builder.toml can just be ignored.

Reply:
Yeah, I think stacks can be ignored, and we will follow the steps described in the [plan](https://docs.google.com/document/d/1bExpr31U5R5yQ6fncpl5YcdosWVwYcXgkt12vE-lpvU/edit#heading=h.jpmej4uevvh8) to keep compatibility